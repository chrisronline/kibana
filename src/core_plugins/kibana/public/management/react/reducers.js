/* eslint-disable */
import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux'

import app from './app/app.reducer';
import indexPattern from './index-pattern/index-pattern.reducer';

export const rootReducer = combineReducers({
  routing: routerReducer,
  app,
  indexPattern,
});
