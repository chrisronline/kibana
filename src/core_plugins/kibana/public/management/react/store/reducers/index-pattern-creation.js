import { handleActions } from 'redux-actions';
import { createSelector } from 'reselect';
import { chunk, sortBy as sortByLodash } from 'lodash';

import {
  selectTimeField,
  fetchedTimeFields,
  goToNextPage,
  goToPreviousPage,
  changeSort,
  fetchedIndices,
  includeSystemIndices,
  excludeSystemIndices,
  creatingIndexPattern,
  createdIndexPattern,
} from '../actions/index-pattern-creation';

import {
  getIndexPatternCreate,
} from '../../reducers';

const defaultState = {
  isIncludingSystemIndices: false,
  isCreating: false,
  timeFields: {
    timeFields: undefined,
    selectedTimeField: undefined,
  },
  results: {
    indices: undefined,
    pattern: undefined,
    hasExactMatches: false,
    perPage: 10,
    page: 1,
    sortBy: undefined,
    sortAsc: true,
  }
}

export default handleActions({
  [selectTimeField](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [fetchedTimeFields](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [selectTimeField](state, { payload }) {
    const { timeField } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        selectedTimeField: timeField,
      },
    };
  },
  [fetchedIndices](state, { payload }) {
    const { indices, hasExactMatches, pattern } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        indices,
        hasExactMatches,
        pattern,
      },
    };
  },
  [changeSort](state, { payload }) {
    const { sortBy, sortAsc } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        sortBy,
        sortAsc: sortBy === state.results.sortBy ? !state.results.sortAsc: sortAsc,
      },
    };
  },
  [goToNextPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getNextPage(state.results),
      },
    };
  },
  [goToPreviousPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getPreviousPage(state.results),
      },
    };
  },
  [includeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: true,
    };
  },
  [excludeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: false,
    };
  },
  [creatingIndexPattern](state, action) {
    return {
      ...state,
      isCreating: true,
    };
  },
  [createdIndexPattern](state, action) {
    return {
      ...state,
      isCreating: false,
    };
  },
}, defaultState);

function getNextPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const nextPage = page + 1;
  return nextPage > pages
    ? 1
    : nextPage;
}

function getPreviousPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const previousPage = page - 1;
  return previousPage < 1
    ? pages
    : previousPage;
}

function getFilteredIndices(indices, isIncludingSystemIndices) {
  if (!indices) {
    return indices;
  }

  if (isIncludingSystemIndices) {
    return indices;
  }

  // All system indices begin with a period.
  return indices.filter(index => !index.name.startsWith('.'));
}

function getPaginatedIndices(indices, page, perPage) {
  const pagesOfIndices = chunk(indices, perPage);
  return pagesOfIndices[page - 1];
}

export const getFilteredAndPaginatedIndices = createSelector(
  [
    state => state.results,
    state => state.isIncludingSystemIndices
  ],
  ({
    indices: allIndices,
    page,
    perPage,
    sortBy,
    sortAsc,
  }, isIncludingSystemIndices) => {
    if (allIndices === undefined) {
      return {
        indices: undefined,
        numOfPages: 0,
      };
    }

    let filteredIndices = getFilteredIndices(allIndices, isIncludingSystemIndices);
    if (!!sortBy) {
      filteredIndices = sortByLodash(filteredIndices, sortBy);
      if (!sortAsc) {
        filteredIndices.reverse();
      }
    }

    const numOfPages = Math.ceil(filteredIndices.length / perPage);
    const indices = getPaginatedIndices(filteredIndices, page, perPage);
    return {
      indices,
      numOfPages,
    }
  }
);

export const getPattern = state => getIndexPatternCreate(state).results.pattern;
export const getSelectedTimeField = state => getIndexPatternCreate(state).timeFields.selectedTimeField;
