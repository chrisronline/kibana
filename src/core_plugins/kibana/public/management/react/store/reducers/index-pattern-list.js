import { handleActions } from 'redux-actions';

import {
  fetchedIndexPatterns,
  setTransientTableId,
} from '../actions/index-pattern-list';

import {
  getIndexPatternList,
} from '../../reducers';

const defaultState = {
  indexPatterns: undefined,
  transientTableId: undefined,
};

export default handleActions({
  [fetchedIndexPatterns](state, { payload }) {
    const { indexPatterns } = payload;

    return {
      ...state,
      indexPatterns,
    };
  },
  [setTransientTableId](state, { payload: { id: transientTableId } }) {
    return {
      ...state,
      transientTableId,
    };
  }
}, defaultState);

export const getIndexPatterns = state => getIndexPatternList(state).indexPatterns;
export const getPathToIndexPatterns = () => 'indexPatterns';
