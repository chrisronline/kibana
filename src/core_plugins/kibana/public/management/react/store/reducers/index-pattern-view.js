import { handleActions } from 'redux-actions';

import {
  fetchedIndexPattern,
  setAsDefaultIndexPattern,
} from '../actions/index-pattern-view';

import {
  getIndexPatternView,
} from '../../reducers';

const defaultState = {
  indexPattern: {
    pattern: undefined,
    fields: undefined,
    timeFieldName: undefined,
    isDefault: undefined,
    id: undefined,
  },
  fieldsTable: {},
  tabs: {},
};

export const indexPatternView = handleActions({
  [fetchedIndexPattern](state, { payload: { indexPattern } }) {
    return {
      ...state,
      indexPattern,
    };
  },
  [setAsDefaultIndexPattern](state) {
    return {
      ...state,
      indexPattern: {
        ...state.indexPattern,
        isDefault: true,
      }
    };
  }
}, defaultState);

export const getPathToFields = () => 'indexPattern.fields';
export const getTabs = state => getIndexPatternView(state).tabs;
