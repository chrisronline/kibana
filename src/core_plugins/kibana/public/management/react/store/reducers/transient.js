import { handleActions } from 'redux-actions';

import {
  change,
  createItem,
} from '../actions/transient';

const defaultState = {};

export default handleActions({
  [createItem](state, { payload: { id, defaultData }}) {
    return {
      ...state,
      [id]: defaultData || {},
    }
  },
  [change](state, { payload: { id, data } }) {
    return {
      ...state,
      [id]: {
        ...state[id],
        ...data
      },
    };
  },
}, defaultState);
