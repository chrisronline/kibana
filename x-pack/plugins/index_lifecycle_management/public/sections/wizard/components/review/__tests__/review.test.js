/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { Review } from '../';
import { indexLifecycleManagement } from '../../../../../store/reducers';
import { STRUCTURE_POLICY_NAME } from '../../../../../store/constants';

jest.useFakeTimers();

jest.mock('@elastic/eui', () => ({
  EuiTitle: props => props.children || '',
  EuiCode: props => props.children || '',
  EuiCallOut: props => props.children || '',
  EuiSpacer: props => props.children || '',
  EuiHorizontalRule: props => props.children || '',
  EuiButton: props => props.children || '',
  EuiFlexItem: props => props.children || '',
  EuiText: props => props.children || '',
  EuiButtonEmpty: props => props.children || '',
  EuiFormRow: props => props.children || '',
  EuiSwitch: props => props.children || '',
  EuiFieldText: props => props.children || '',
  EuiLoadingSpinner: props => props.children || '',
  EuiFlexGroup: props => props.children || '',
}));

jest.mock('brace/theme/github', () => {});
jest.mock('brace/mode/json', () => {});
jest.mock('brace/snippets/json', () => {});
jest.mock('brace/ext/language_tools', () => {});

jest.mock('../diff_view', () => ({
  DiffView: () => ''
}));

jest.mock('../../../../../store/actions', () => ({
  setSelectedPolicyName: jest.fn(),
  setSaveAsNewPolicy: jest.fn(),
}));

jest.mock('../../../../../api', () => ({
  getAffectedIndices: () => ([
    'foobar*',
    'boofar*'
  ]),
}));

const defaultState = indexLifecycleManagement({}, {});
const defaultStore = {
  subscribe: jest.fn(),
  dispatch: jest.fn(),
};
const actions = require('../../../../../store/actions');

describe('Review', () => {
  it('should render properly', async () => {
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}],
          fullSelectedIndexTemplate: {
            settings: {
              index: {
                number_of_shards: '' + defaultState.nodes.selectedPrimaryShardCount,
                number_of_replicas: '' + defaultState.nodes.selectedReplicaCount,
                lifecycle: {
                  name: defaultState.policies.selectedPolicy.name,
                },
                routing: {
                  allocation: {
                    include: {
                      sattr_name: defaultState.nodes.selectedNodeAttrs,
                    }
                  }
                }
              }
            }
          }
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <Review
          validate={() => {}}
          done={() => {}}
          errors={{
            [STRUCTURE_POLICY_NAME]: []
          }}
          back={() => {}}
        />
      </Provider>
    );

    jest.runAllTimers();
    // Ensure all promises resolve
    await new Promise(resolve => process.nextTick(resolve));
    // Ensure the state changes are reflected
    component.update();

    expect(component.find('Review')).toMatchSnapshot();
  });

  it('should a loading state for affected indices', () => {
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <Review
          validate={() => {}}
          done={() => {}}
          errors={{
            [STRUCTURE_POLICY_NAME]: []
          }}
          back={() => {}}
        />
      </Provider>
    );

    expect(component.find('EuiLoadingSpinner').length).toBe(1);
  });

  it('should show the aliases if bootstrap is enabled', () => {
    const store = {
      getState: () => ({
        ...defaultState,
        general: {
          ...defaultState.general,
          bootstrapEnabled: true,
          aliasName: 'my_alias'
        },
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <Review
          validate={() => {}}
          done={() => {}}
          errors={{
            [STRUCTURE_POLICY_NAME]: []
          }}
          back={() => {}}
        />
      </Provider>
    );

    expect(component.find('EuiCallOut').get(1)).toMatchSnapshot();
  });

  it('should show the diff if there are changes', () => {
    const store = {
      getState: () => ({
        ...defaultState,
        general: {
          ...defaultState.general,
          bootstrapEnabled: true,
          aliasName: 'my_alias'
        },
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}],
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <Review
          validate={() => {}}
          done={() => {}}
          errors={{
            [STRUCTURE_POLICY_NAME]: []
          }}
          back={() => {}}
        />
      </Provider>
    );

    expect(component.find('DiffView')).toMatchSnapshot();
  });

  it('should show a warning if editing an existing policy', () => {
    const store = {
      getState: () => ({
        ...defaultState,
        policies: {
          ...defaultState.policies,
          originalPolicyName: 'foo',
          selectedPolicy: {
            ...defaultState.policies.selectedPolicy,
            saveAsNew: false
          }
        },
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}],
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <Review
          validate={() => {}}
          done={() => {}}
          errors={{
            [STRUCTURE_POLICY_NAME]: []
          }}
          back={() => {}}
        />
      </Provider>
    );

    expect(component.find('EuiText')).toMatchSnapshot();
  });

  it('should finish the wizard when clicking the submit button', () => {
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    const done = jest.fn();
    const component = mount(
      <Provider store={store}>
        <Review
          validate={() => {}}
          done={done}
          errors={{
            [STRUCTURE_POLICY_NAME]: []
          }}
          back={() => {}}
        />
      </Provider>
    );

    component.find('Review').find('EuiButton').get(0).props.onClick();
    expect(done).toHaveBeenCalled();
  });

  it('should show errors', () => {
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    const done = jest.fn();
    const component = mount(
      <Provider store={store}>
        <Review
          validate={() => {}}
          done={done}
          errors={{
            [STRUCTURE_POLICY_NAME]: ['Error']
          }}
          back={() => {}}
        />
      </Provider>
    );

    expect(component.find('ErrableFormRow')).toMatchSnapshot();
  });

  describe('should validate fields when they change', () => {
    const store = {
      getState: () => ({
        ...defaultState,
        policies: {
          ...defaultState.policies,
          originalPolicyName: 'foo',
          selectedPolicy: {
            ...defaultState.policies.selectedPolicy,
            saveAsNew: true
          }
        },
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    test('save as new policy', async () => {
      const validate = jest.fn();
      const component = mount(
        <Provider store={store}>
          <Review
            validate={validate}
            done={() => {}}
            errors={{
              [STRUCTURE_POLICY_NAME]: []
            }}
            back={() => {}}
          />
        </Provider>
      );

      await component.find('EuiSwitch').get(0).props.onChange({ target: { checked: true } });
      expect(validate).toHaveBeenCalled();
      expect(actions.setSaveAsNewPolicy).toHaveBeenCalledWith(true);
    });

    test('policy name', async () => {
      const validate = jest.fn();
      const component = mount(
        <Provider store={store}>
          <Review
            validate={validate}
            done={() => {}}
            errors={{
              [STRUCTURE_POLICY_NAME]: []
            }}
            back={() => {}}
          />
        </Provider>
      );

      await component.find('EuiFieldText').get(0).props.onChange({ target: { value: 'lala' } });
      expect(validate).toHaveBeenCalled();
      expect(actions.setSelectedPolicyName).toHaveBeenCalledWith('lala');
    });
  });
});
