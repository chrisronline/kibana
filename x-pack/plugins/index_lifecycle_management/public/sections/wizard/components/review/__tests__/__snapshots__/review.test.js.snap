// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review should render properly 1`] = `
<Review
  affectedIndexTemplates={
    Array [
      "",
    ]
  }
  aliasName=""
  back={[Function]}
  bootstrapEnabled={false}
  done={[Function]}
  errors={
    Object {
      "policyName": Array [],
    }
  }
  indexName=""
  lifecycle={
    Object {
      "name": "",
      "phases": Object {
        "hot": Object {
          "actions": Object {
            "rollover": Object {
              "alias": "",
            },
          },
          "after": "0s",
        },
      },
    }
  }
  policies={Array []}
  saveAsNewPolicy={true}
  selectedIndexTemplateName=""
  selectedPolicyName=""
  setSaveAsNewPolicy={[Function]}
  setSelectedPolicyName={[Function]}
  templateDiff={
    Object {
      "hasChanged": false,
      "newFullIndexTemplate": Object {
        "settings": Object {
          "index": Object {
            "lifecycle": Object {
              "name": "",
            },
            "number_of_replicas": "1",
            "number_of_shards": "1",
            "routing": Object {
              "allocation": Object {
                "include": Object {
                  "sattr_name": "",
                },
              },
            },
          },
        },
      },
      "originalFullIndexTemplate": Object {
        "settings": Object {
          "index": Object {
            "lifecycle": Object {
              "name": "",
            },
            "number_of_replicas": "1",
            "number_of_shards": "1",
            "routing": Object {
              "allocation": Object {
                "include": Object {
                  "sattr_name": "",
                },
              },
            },
          },
        },
      },
    }
  }
  validate={[Function]}
>
  <div
    className="euiAnimateContentLoad"
  >
    <EuiSpacer />
    <EuiTitle
      size="s"
    >
      <h3>
        Review your policy changes
      </h3>
    </EuiTitle>
    <EuiText>
      <p>
        When you save a policy, your changes go into effect immediately.
      </p>
    </EuiText>
    <EuiSpacer />
    <EuiCallOut
      color="warning"
      title="Your changes affect these template configurations"
    >
      <h4>
        1 Index template
      </h4>
      <ul>
        <li />
      </ul>
      <h4>
        2 Indices
      </h4>
      <ul>
        <li
          key="foobar*"
        >
          foobar*
        </li>
        <li
          key="boofar*"
        >
          boofar*
        </li>
      </ul>
    </EuiCallOut>
    <EuiHorizontalRule
      className="ilmHrule"
    />
    <EuiSpacer />
    <EuiTitle
      size="s"
    >
      <h3>
        Save your work
      </h3>
    </EuiTitle>
    <EuiSpacer />
    <ErrableFormRow
      errorKey="policyName"
      errors={
        Object {
          "policyName": Array [],
        }
      }
      label="Policy name"
    >
      <EuiFormRow
        error={Array []}
        label="Policy name"
      >
        <EuiFieldText
          key=".0"
          onChange={[Function]}
          value=""
        />
      </EuiFormRow>
    </ErrableFormRow>
    <EuiFlexGroup>
      <EuiFlexItem
        grow={false}
      >
        <EuiButtonEmpty
          iconSide="left"
          iconType="sortLeft"
          onClick={[Function]}
        >
          Back
        </EuiButtonEmpty>
      </EuiFlexItem>
      <EuiFlexItem
        grow={false}
      >
        <EuiButton
          color="secondary"
          fill={true}
          iconType="check"
          onClick={[Function]}
        >
          Looks good, save changes
        </EuiButton>
      </EuiFlexItem>
    </EuiFlexGroup>
      
  </div>
</Review>
`;

exports[`Review should show a warning if editing an existing policy 1`] = `
Array [
  <EuiText>
    <p>
      When you save a policy, your changes go into effect immediately.
    </p>
  </EuiText>,
  <EuiText>
    <p>
      <strong>
        You are editing an existing policy
      </strong>
      . Any changes you make will also change index templates that this policy is attached to. Alternately, you can save these changes in a new policy and only change the template you selected.
    </p>
  </EuiText>,
]
`;

exports[`Review should show errors 1`] = `
<ErrableFormRow
  errorKey="policyName"
  errors={
    Object {
      "policyName": Array [
        "Error",
      ],
    }
  }
  label="Policy name"
>
  <EuiFormRow
    error={
      Array [
        "Error",
      ]
    }
    label="Policy name"
  >
    <EuiFieldText
      key=".0"
      onChange={[Function]}
      value=""
    />
  </EuiFormRow>
</ErrableFormRow>
`;

exports[`Review should show the aliases if bootstrap is enabled 1`] = `
<EuiCallOut
  color="success"
  title="This action creates a new index alias"
>
  <p>
    You decided to bootstrap a new index. Point to this new alias going forward.
  </p>
  <h3>
    <span
      className="ilmAlias"
    >
      my_alias
    </span>
     is your new alias
  </h3>
</EuiCallOut>
`;

exports[`Review should show the diff if there are changes 1`] = `
<DiffView
  templateDiff={
    Object {
      "hasChanged": true,
      "newFullIndexTemplate": Object {
        "settings": Object {
          "index": Object {
            "lifecycle": Object {
              "name": "",
            },
            "number_of_replicas": "1",
            "number_of_shards": "1",
            "routing": Object {
              "allocation": Object {
                "include": Object {
                  "sattr_name": "",
                },
              },
            },
          },
        },
      },
      "originalFullIndexTemplate": Object {
        "settings": Object {},
      },
    }
  }
/>
`;
